ospf6d/ospf6_intra.c

```diff
@@ -409,105 +471,113 @@ void ospf6_router_lsa_originate(struct event *thread)
+	 * away helps convergence with inter-area routes.
 	 */
 	if (count && !link_state_id)
 		ospf6_spf_schedule(oa->ospf6,
 				   OSPF6_SPF_FLAGS_ROUTER_LSA_ORIGINATED);
 }
 
+void ospf6_router_lsa_originate(struct event *thread)
+{
+	struct ospf6_area *oa = EVENT_ARG(thread);
+
+	router_lsa_originate(oa, OSPF6_LSTYPE_ROUTER);
+	router_lsa_originate(oa, OSPF6_LSTYPE_E_ROUTER);
+}
+
 /*******************************/
 /* RFC2740 3.4.3.2 Network-LSA */
 /*******************************/
 
+/* FIXME: this doesn't have topotest coverage */
 static char *ospf6_network_lsa_get_ar_id(struct ospf6_lsa *lsa, char *buf,
 					 int buflen, int pos)
 {
-	char *start, *end, *current;
-	struct ospf6_network_lsa *network_lsa;
-	struct ospf6_network_lsdesc *lsdesc;
+	struct ospf6_network_lsdesc *lsdesc = nth_lsdesc(lsa->header, pos);
 
-	if (lsa) {
-		network_lsa = (struct ospf6_network_lsa
-				       *)((caddr_t)lsa->header
-					  + sizeof(struct ospf6_lsa_header));
-
-		start = (char *)network_lsa + sizeof(struct ospf6_network_lsa);
-		end = (char *)lsa->header + ntohs(lsa->header->length);
-		current = start + pos * (sizeof(struct ospf6_network_lsdesc));
-
-		if ((current + sizeof(struct ospf6_network_lsdesc)) <= end) {
-			lsdesc = (struct ospf6_network_lsdesc *)current;
-			if (buf) {
-				inet_ntop(AF_INET, &lsdesc->router_id, buf,
-					  buflen);
-				return buf;
-			}
-		}
-	}
+	if (!lsdesc || !buf || buflen < (1 + INET_ADDRSTRLEN))
+		return NULL;
 
-	return NULL;
+	inet_ntop(AF_INET, &lsdesc->router_id, buf, buflen);
+	return buf;
 }
 
+/* FIXME: this doesn't have topotest coverage */
+static char *ospf6_e_network_lsa_get_ar_id(struct ospf6_lsa *lsa, char *buf,
+					 int buflen, int pos)
+{
+	struct tlv_attached_routers *tlv = nth_tlv(lsa->header, pos);
+
+	if (!tlv || !buf || buflen < (1 + INET_ADDRSTRLEN))
+		return NULL;
+
+	inet_ntop(AF_INET, &tlv->router_id, buf, buflen);
+	return buf;
+}
+
+static int cb_print_network_lsdesc(void *desc, void *cb_data)
+{
+	struct ospf6_network_lsdesc *lsdesc = desc;
+	struct cbd_lsdesc_printer *cbd = cb_data;
+	char buf[128];
+
+	inet_ntop(AF_INET, &lsdesc->router_id, buf, sizeof(buf));
+	if (cbd->use_json)
+		json_object_array_add(cbd->json_arr,
+				      json_object_new_string(buf));
+	else
+		vty_out(cbd->vty, "     Attached Router: %s\n", buf);
+	return 0;
+}
+
+
 static int ospf6_network_lsa_show(struct vty *vty, struct ospf6_lsa *lsa,
 				  json_object *json_obj, bool use_json)
 {
-	char *start, *end, *current;
 	struct ospf6_network_lsa *network_lsa;
-	struct ospf6_network_lsdesc *lsdesc;
-	char buf[128], options[32];
-	json_object *json_arr = NULL;
-
-	network_lsa =
-		(struct ospf6_network_lsa *)((caddr_t)lsa->header
-					     + sizeof(struct ospf6_lsa_header));
+	char options[32];
+	struct cbd_lsdesc_printer cbd = { .vty = vty, .use_json = use_json };
+	struct tlv_handler h1 = { .tlv_type = OSPF6_TLV_RESERVED,
+				  .callback = cb_print_network_lsdesc,
```

Comment:
Function pointers on the stack were marked as a vulnerable anti-pattern in https://github.com/FRRouting/frr/pull/16199#issuecomment-2219751707